### PR TITLE
K/N: Track safe points in the aggressive GC

### PR DESCRIPTION
**Pull request by ilmat192**

### CODE REVIEW

File: kotlin-native/backend.native/tests/objcexport/values.swift [hunk 1]
```diff
@@ -896,9 +896,15 @@ func testWeakRefs0(frozen: Bool) throws {
             Holder.ref2 = obj2
         }
 
+// With aggressive GC, the objects referenced from the Holder are freed once
+// we left the autoreleasepool scope, before ValuesKt.gc().
+// TODO: This behaviour is possible for the regular GC too (but much less probable),
+// so we need to handle it somehow. E.g. by disabling GC at all during setting up the test.
```
[the remark this answers is not in the document]
Comment (ilmat192) [reply kotlin-native/backend.native/tests/objcexport/values.swift]: Well, it seems that we have no such checks in other tests, so ok, let's remove them.

File: kotlin-native/runtime/src/gc/cms/cpp/FinalizerProcessor.cpp [hunk 2]
```diff
@@ -10,6 +10,9 @@
 
 void kotlin::gc::FinalizerProcessor::StartFinalizerThreadIfNone() noexcept {
     if (finalizerThread_.joinable()) return;
```
[the remark this answers is not in the document]
Comment (ilmat192) [reply kotlin-native/runtime/src/gc/cms/cpp/FinalizerProcessor.cpp]: Ok, removed

File: kotlin-native/runtime/src/gc/common/cpp/GCSchedulerImpl.hpp [hunk 3]
```diff
@@ -79,6 +81,41 @@ class RegularIntervalPacer {
     std::atomic<TimePoint> lastGC_;
 };
 
+template <size_t SafePointStackSize = 16>
+class SafePointTracker {
+public:
+    using SafePointID = kotlin::StackTrace<SafePointStackSize>;
+
+    SafePointTracker(size_t maxSize = 100000) : maxSize_(maxSize) {}
```
[the remark this answers is not in the document]
Comment (ilmat192) [reply kotlin-native/runtime/src/gc/common/cpp/GCSchedulerImpl.hpp]: Thank you, added.

**Comment by ilmat192** [comment]
Sorry for the delay, for some reason I didn't receive an e-mail notification about the comments :( I've added the fixes. `hostRuntimeTests` and backend.native `sanity` are green (at least on MacOS).

**Comment by projedi** [comment]
> Sorry for the delay, for some reason I didn't receive an e-mail notification about the comments :( I've added the fixes. `hostRuntimeTests` and backend.native `sanity` are green (at least on MacOS).

Cool, thanks! They're green on CI too after a rebase. Though, the linux build now takes 3 hours instead of 40 minutes. We'll need to disable some tests to have it integrated, I think. I'll look into it, but did you have an idea of potential tests that we can disable?

**Comment by ilmat192** [comment]
Rebaed onto master and fixed includes on `GCSchedulerImpl.hpp`

> I'll look into it, but did you have an idea of potential tests that we can disable?

Thanks! I've collected a [build scan](https://scans.gradle.com/s/4hc7qagmw3gua/timeline?sort=longest) for stanity tests on MacOS and there are several long tasks that look like good candidates for disabling in the aggressive mode:

 - `stress_gc_allocations`
 - `harmonyRegexTest`
 - `debugger_test`
 - several tasks testing the `konanc` shell script (tasks with type `KonanDriverTest`).

I do not remember if we run some stdlib tests other than harmony ones in aggressive precommit CI runs. If yes, we can try disabling them too.

**Pull request closed without merging**